[BLOG] Neural Sparse is now available in HuggingFace SentenceTransformers (#3799)
opensearch-project/project-website: _posts/2025-06-11-neural-sparse-sentence-transformers.md

  - zhichaog
  - yych
  - tomaarsen
  - arthurbr11
date: 2025-06-11
categories:
    - technical-posts
has_science_table: true
meta_keywords: neural sparse models, OpenSearch semantic search, Sentence Transformers
meta_description: Using OpenSearch neural sparse features with Sentence Transformers
excerpt: Through a collaborative effort between the OpenSearch and Hugging Face teams, neural sparse is now available in the Sentence Transformers library.
featured_blog_post: false 
featured_image: false # /assets/media/blog-images/__example__image__name.jpg
---

OpenSearch's neural sparse feature transforms text into sparse token-weight pairs using transformer models, combining semantic search capabilities with efficient inverted indexing. This approach delivers high retrieval accuracy with low latency and minimal resource usage, offering a scalable alternative to dense vector search. The inference-free mode eliminates query encoding overhead, enabling extremely fast search. OpenSearch provides state-of-the-art pre-trained models that consistently lead the [BEIR](https://github.com/beir-cellar/beir) benchmark for sparse retrievers, and has yield over **10M+ downloads** on Hugging Face!

> The [Sentence Transformers](https://huggingface.co/sentence-transformers)(a.k.a. SBERT) library, developed by [UKPLab](https://www.tu-darmstadt.de/) and maintained by [Hugging Face](https://huggingface.co/), is a Python framework designed to generate semantically meaningful embeddings for sentences, paragraphs, and images. The library is now widely adopted in the natural language processing (NLP) tasks such as semantic search, clustering, paraphrase mining, and retrieval-augmented generation.

Author: ack

<img src="/assets/media/blog-images/2025-06-11-neural-sparse-sentence-transformers/image.png"/>

## Neural Sparse models in Sentence Transformers

Compared to dense embedding models, sparse models have a more complex structure, requiring post-processing layers for vector sparsification and complex tensor-to-map conversions for OpenSearch compatibility. These factors collectively increase the technical barrier to using and deploying neural sparse models in Python.

Sentence Transformers offers a streamlined user experience. All encoding details of OpenSearch neural sparse models are abstracted away. Configuration aspects like bi-encoder or doc-only (inference-free) mode and activation functions are handled internally. Model initialization and text encoding require just 1-2 lines of code.

Now let's explore neural sparse in Sentence Transformers with hands-on coding examples!

### Install python dependencies
```

> pip install sentence-transformers==4.2.0

Author: ack

```

### Model Initialization

See OpenSearch's [Hugging Face repo](https://huggingface.co/opensearch-project) for all available model ids.

```python
# Here we initialize a model directly from model_id.
# All settings like inference-free and l0 activation are bind with model_id and handled silently.
from sentence_transformers import SparseEncoder

sparse_model = SparseEncoder("opensearch-project/opensearch-neural-sparse-encoding-doc-v3-distill")
```

### Encoding Examples

The model supports encoding for both documents and queries.

#### Query Encoding
```python
# We first use encode method to convert text to torch.sparse_coo tensor (it's inference-free here)
# Then we use decode method to convert torch.sparse_coo tensor to (token, weight) pairs

> query_tensor = sparse_model.encode([{"query": "What's the weather in ny now?"}])

Author: Yes let's update it after merge

<img src="/assets/media/blog-images/2025-06-11-neural-sparse-sentence-transformers/image.png"/>

## How to use neural sparse models in Sentence Transformers

Compared to dense embedding models, sparse models have a more complex structure, requiring post-processing layers for vector sparsification and complex tensor-to-map conversions for OpenSearch compatibility. These factors collectively increase the technical barrier to using and deploying neural sparse models in Python.

Sentence Transformers offers a streamlined user experience to simplify this process. It abstracts the encoding details of OpenSearch neural sparse models, including configuration options like bi-encoder or doc-only (inference-free) modes and activation functions. These abstractions are handled internally, so you can initialize a model and encode text with just one or two lines of code.

Follow these steps to get started with neural sparse models using the Sentence Transformers library.

### Step 1: Install Python dependencies

First, install the required Python library:

```bash
pip install sentence-transformers==5.0.0
```

### Step 2: Initialize a neural sparse model

> Import the SparseEncoder class and load a pretrained model. You can find more models in OpenSearch's [Hugging Face repo](https://huggingface.co/opensearch-project):

Reviewer: ```suggestion
Import the SparseEncoder class and load a pretrained model. You can find more models in OpenSearch's [Hugging Face repository](https://huggingface.co/opensearch-project):
```

    'ny': 0.9702811241149902,
    'rain': 0.9549840092658997,
    'rainy': 0.9479081630706787,
    'nyc': 0.8448441624641418,
    'new': 0.687921404838562,
    'raining': 0.6790128350257874,
    'current': 0.6762210130691528,
    'wet': 0.6448971629142761
}
```
### Step 5: Integrate with OpenSearch

Once you obtained the sparse embeddings from Sentence Transformers, you can index and search them directly in OpenSearch. The output format is fully compatible with OpenSearch's neural sparse search.

For detailed instructions, see the [OpenSearch documentation](https://docs.opensearch.org/docs/latest/vector-search/ai-search/neural-sparse-with-raw-vectors/).

You can also explore example scripts for working with the following embedding types:

- [Sparse embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sparse_encoder/applications/semantic_search/semantic_search_opensearch.py)
- [Dense embeddings](https://github.com/UKPLab/sentence-transformers/blob/master/examples/sentence_transformer/applications/semantic-search/semantic_search_nq_opensearch.py)

## OpenSearch's contributions to inference-free models in Sentence Transformers

> The OpenSearch team is leading innovation in learned sparse retrieval. Our research in inference-free sparse encoders [^1][^2] is increasingly adopted as standard practice across the industry. The following techniques have now been integrated into the Sentence Transformers library, making them more accessible to the broader community.

Reviewer: ```suggestion
OpenSearch is driving innovation in learned sparse retrieval. Our research into inference-free sparse encoders [^1][^2] is increasingly adopted as standard practice across the industry. The following techniques have now been integrated into the Sentence Transformers library, making them more accessible to the broader community.
```

> 

Author: Our intention is to refer to OpenSearch. Because those 2 models are trained by OpenSearch and released at Hugging Face. Let me change it to OpenSearch.

### IDF-aware Penalty: Enhancing Relevance and Efficiency

Traditional sparse retrievers apply uniform penalties to all tokens during training, ignoring the varying importance of different terms. To address this limitation, OpenSearch researchers introduced IDF-aware penalty[^1], which modulates the penalty based on token significance. This approach applies stronger penalties to common tokens while preserving rare, information-rich tokens that carry more distinctive meaning.

> 

Reviewer: Line 142: Either "an" or "the" should precede "IDF-aware penalty". The 2nd sentence reads as repetitive. An/the IDF-aware penalty modulates the penalty? Can we clarify?

> With this enhancement, inference-free retrievers significantly outperform baselines on zero-shot retrieval. The IDF-aware penalty also improves search efficiency by reducing average FLOPs on inverted indexes. Importantly, this enhancement serves as a prerequisite for implementing the powerful [two-phase search](https://opensearch.org/blog/Introducing-a-neural-sparse-two-phase-algorithm) methodology. All inference-free models[^3] released by our team incorporate this enhancement.

Reviewer: Last sentence: "our team" => "Hugging Face"?

Author: it's OpenSearch. I'll change it

The IDF-aware enhancement is implemented as a [module](https://github.com/UKPLab/sentence-transformers/blob/v5.0-release/sentence_transformers/sparse_encoder/models/SparseStaticEmbedding.py) in Sentence Transformers. You can create inference-free sparse models with IDF-aware enhancement by incorporating this module into your model architecture. The implementation requires configuring a document encoder and router setup, as shown in the following code example:

```python
from sentence_transformers import SparseEncoder, models
from sentence_transformers.sparse_encoder.models import SparseStaticEmbedding, MLMTransformer, SpladePooling

doc_encoder = MLMTransformer(model_id)
router = models.Router(
    {
        "query": [
            SparseStaticEmbedding.from_json(
                model_id,
                tokenizer=doc_encoder.tokenizer,
                frozen=True,
            ),
        ],
        "document": [
            doc_encoder,
            SpladePooling(),
        ],
    }
)

sparse_model = SparseEncoder(
    modules=[router],
    similarity_fn_name="dot",
)
```

### ℓ0-inspired sparsification: Advancing sparse retrieval

> Achieving optimal balance between sparsity and relevance has been a persistent challenge for inference-free retrievers. The OpenSearch team addressed this by developing two novel ℓ0-inspired sparsification approaches[^2] (accepted to SIGIR 2025): ℓ0 mask loss, which excludes already-sparse documents from further regularization, and ℓ0 Approximation Activation, which uses log transformations to alter the penalty on tokens with lower activations.

Reviewer: ```suggestion
Achieving optimal balance between sparsity and relevance has been a persistent challenge for inference-free retrievers. OpenSearch addressed this by developing two novel ℓ0-inspired sparsification approaches[^2] (accepted to SIGIR 2025): ℓ0 mask loss, which excludes already-sparse documents from further regularization, and ℓ0 Approximation Activation, which uses log transformations to alter the penalty on tokens with lower activations.
```

Models using these ℓ0-inspired techniques demonstrate superior performance while significantly improving efficiency and sparsity. The ℓ0 mask loss also streamlines the training process by eliminating the need for multiple attempts at tuning the FLOPS lambda hyperparameter.

> Our [doc-v3-distill](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-distill) and [doc-v3-gte](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-gte) models incorporate these enhancements, delivering state-of-the-art performance.

Reviewer: ```suggestion
Hugging Face's [doc-v3-distill](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-distill) and [doc-v3-gte](https://huggingface.co/opensearch-project/opensearch-neural-sparse-encoding-doc-v3-gte) models incorporate these enhancements, delivering state-of-the-art performance.
```

```python
# ℓ0 Mask Loss use threshold 150
from sentence_transformers.sparse_encoder.losses import FlopsLoss
loss = FlopsLoss(model, threshold=150)

# ℓ0 Approximation Activation
from sentence_transformers.sparse_encoder.models import SpladePooling
pooling = SpladePooling(activation_function="log1p_relu")
```

## Further reading

For more information about neural sparse search, see these previous blog posts:

- [Improving document retrieval with sparse semantic encoders](https://opensearch.org/blog/improving-document-retrieval-with-sparse-semantic-encoders)
- [A deep dive into faster semantic sparse retrieval in OpenSearch 2.12](https://opensearch.org/blog/A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12/)
- [Introducing the neural sparse two-phase algorithm](https://opensearch.org/blog/Introducing-a-neural-sparse-two-phase-algorithm)

> - [Advancing Search Quality and Inference Speed with v2 Series Neural Sparse Models](https://opensearch.org/blog/neural-sparse-v2-models)

Reviewer: ```suggestion
- [Advancing search quality and inference speed with v2 series neural sparse models](https://opensearch.org/blog/neural-sparse-v2-models)
```
